core/packages/contracts/src/InboundChannel.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/utils/cryptography/MerkleProof.sol";
+import "openzeppelin/access/AccessControl.sol";
+import "./ParachainClient.sol";
+import "./IRecipient.sol";
+import "./IVault.sol";
+
+contract InboundChannel is AccessControl {
+    // Nonce for each origin
+    mapping(bytes => uint64) public nonce;
+
+    // Registered message handlers
+    mapping(uint16 => Handler) handlers;
+
+    // Light client message verifier
+    IParachainClient public parachainClient;
+
+    // Relayers are rewarded from this vault
+    IVault public vault;
+
+    // The relayer reward for submitting a message
+    uint256 public reward;
+
+    // The governance contract, which is a proxy for Polkadot governance, administers via this role
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+
+    // Relayers must provide enough gas to cover message dispatch plus this buffer
+    uint256 public constant GAS_BUFFER = 24000;
+
+    // Inbound message from BridgeHub parachain
+    struct Message {
+        bytes origin;
+        uint64 nonce;
+        uint16 handler;
+        bytes payload;
+    }
+
+    // Message handler entry
+    struct Handler {
+        // Address of message handler which implements IRecipient
+        address recipient;
+        // Amount of gas to forward to message handler
+        uint32 gasToForward;
+    }
+
+    // The result of message dispatch
+    struct DispatchResult {
+        // Whether the dispatch succeeded
+        bool succeeded;
+        // Various error signifiers
+        string errorReason;
+        uint256 errorPanicCode;
+        bytes errorReturnData;
+    }
+
+    event MessageDispatched(bytes origin, uint64 nonce, DispatchResult result);
+    event HandlerUpdated(uint16 id, Handler handler);
+    event ParachainClientUpdated(address parachainClient);
+    event VaultUpdated(address vault);
+    event RewardUpdated(uint256 reward);
+
+    error InvalidProof();
+    error InvalidNonce();
+    error InvalidHandler();
+    error NotEnoughGas();
+
+    constructor(IParachainClient _parachainClient, IVault _vault, uint256 _reward) {
+        _grantRole(ADMIN_ROLE, msg.sender);
```

Comment:
Admin is just deployer. Needs to be changed to Polkadot governance? How will this work? Seems like there are 2 options here, ie, either the InboundChannel becomes an IRecipient too and then calls handle on itself with the Polkadot governance origin and with the update functions checking the origin, or, we have a sovereign account for Polkadot governance deployed onto ethereum that does the calls?